include/dice/template-library/pool_allocator.hpp

```diff
@@ -0,0 +1,205 @@
+	 */
+	template<typename T, size_t ...bucket_sizes>
+	struct pool_allocator {
+		using value_type = T;
+		using pointer = T *;
+		using const_pointer = T const *;
+		using void_pointer = void *;
+		using const_void_pointer = void const *;
+		using size_type = size_t;
+		using difference_type = std::ptrdiff_t;
+
+		using propagate_on_container_copy_assignment = std::true_type;
+		using propagate_on_container_move_assignment = std::true_type;
+		using propagate_on_container_swap = std::true_type;
+		using is_always_equal = std::false_type;
+
+    	template<typename U>
+    	struct rebind {
+    		using other = pool_allocator<U, bucket_sizes...>;
+    	};
+
+	private:
+		template<typename, size_t ...>
+		friend struct pool_allocator;
+
+    	pool<bucket_sizes...> *pool_;
+
+	public:
+		explicit constexpr pool_allocator(pool<bucket_sizes...> &parent_pool) noexcept
+			: pool_{&parent_pool} {
+		}
+
+		constexpr pool_allocator(pool_allocator const &other) noexcept = default;
+		constexpr pool_allocator(pool_allocator &&other) noexcept = default;
+		constexpr pool_allocator &operator=(pool_allocator const &other) noexcept = default;
+		constexpr pool_allocator &operator=(pool_allocator &&other) noexcept = default;
+		constexpr ~pool_allocator() noexcept = default;
+
+		template<typename U>
+		constexpr pool_allocator(pool_allocator<U, bucket_sizes...> const &other) noexcept
+			: pool_{other.pool_} {
+		}
+
+		constexpr pointer allocate(size_t n) {
+			return static_cast<pointer>(pool_->allocate(sizeof(T) * n));
+		}
+
+		constexpr void deallocate(pointer ptr, size_t n) {
+			pool_->deallocate(ptr, sizeof(T) * n);
+		}
+
+		constexpr pool_allocator select_on_container_copy_construction() const {
+			return pool_allocator{*pool_};
+		}
+
+		friend constexpr void swap(pool_allocator &lhs, pool_allocator &rhs) noexcept {
+			using std::swap;
+			swap(lhs.pool_, rhs.pool_);
+		}
+
+		bool operator==(pool_allocator const &other) const noexcept = default;
+		bool operator!=(pool_allocator const &other) const noexcept = default;
+	};
+
+	template<size_t ...bucket_sizes>
+	struct pool {
+		static_assert(sizeof...(bucket_sizes) > 0,
+					  "must at least provide one bucket size, otherwise this would never use a pool for allocation");
+		static_assert(std::ranges::is_sorted(std::array<size_t, sizeof...(bucket_sizes)>{bucket_sizes...}),
+			          "bucket_sizes parameters must be sorted (small to large)");
+
+		using size_type = size_t;
+		using difference_type = std::ptrdiff_t;
+
+		template<typename T>
+		using allocator_type = pool_allocator<T, bucket_sizes...>;
+
+	private:
+		// note: underlying allocator can not be specified via template parameter
+		// because that would be of very limited usefulness, as boost::pool requires the allocation/deallocation functions
+		// to be `static`
+		using pool_type = boost::pool<boost::default_user_allocator_new_delete>;
+
+		std::array<pool_type, sizeof...(bucket_sizes)> pools_;
```

Comment:
I suggest documenting that the capacity of each pool is handled internally by boost.

Reply:
Placed a comment at the first `pool` declaration

Reply:
In the comment you mention the `chunk_size`. I think that we should also mention that the pools' capacity is handled by boost.